### ISSUE TITLE
Could not enable TOTP

### ISSUE THREAD

**Issue by Tthecreator** [nextcloud/twofactor_totp#1112]
### Steps to reproduce
1. install TOTP app
2. go to settings -> security
3. Check the checkmark left of "Enable TOTP" 

### Expected behaviour
TOTP should be enabled

### Actual behaviour
I get a popup message saying "TOTP"

### Server configuration

**Operating system**: Debian/Docker

**Web server:** Apache

**Database:** Mysql 8.0

**PHP version:** 7.4

**Version:** (see admin page) 21.0.1

**Updated from an older version or fresh install:**
I updated from version 18, but thinks got pretty nasty. I updated to 19 correctly and then accidentally skipped to 21. This should not be a problem by definition, but my disk space was full without me realisizing. So therefore my config.php got wiped and I had to go trough a lot of trouble to get things back up and running. I actually couldn't log in and I truncated the oc_storages_credentials table manually in my database, disabled totp for my account and disabled the default encryption module.

**List of activated apps:**
  - accessibility: 1.7.0
  - activity: 2.14.3
  - admin_audit: 1.11.0
  - analytics: 3.4.1
  - bruteforcesettings: 2.1.0
  - calendar: 2.2.1
  - checksum: 1.1.2
  - cloud_federation_api: 1.4.0
  - comments: 1.11.0
  - contacts: 3.5.1
  - contactsinteraction: 1.2.0
  - dashboard: 7.1.0
  - dav: 1.17.1
  - deck: 1.4.2
  - epubreader: 1.4.6
  - federatedfilesharing: 1.11.0
  - federation: 1.11.0
  - files: 1.16.0
  - files_accesscontrol: 1.11.0
  - files_downloadactivity: 1.10.0
  - files_external: 1.12.0
  - files_markdown: 2.3.3
  - files_pdfviewer: 2.1.0
  - files_rightclick: 1.0.0
  - files_sharing: 1.13.1
  - files_texteditor: 2.14.0
  - files_trashbin: 1.11.0
  - files_versions: 1.14.0
  - files_videoplayer: 1.10.0
  - firstrunwizard: 2.10.0
  - fulltextsearch: 21.0.0
  - groupfolders: 9.0.1
  - integration_github: 1.0.0
  - logreader: 2.6.0
  - lookup_server_connector: 1.9.0
  - nextcloud_announcements: 1.10.0
  - notes: 4.0.4
  - notifications: 2.9.0
  - oauth2: 1.9.0
  - password_policy: 1.11.0
  - photos: 1.3.0
  - privacy: 1.5.0
  - provisioning_api: 1.11.0
  - quickaccesssorting: 1.1.2
  - quota_warning: 1.10.0
  - ransomware_protection: 1.10.0
  - recognize: 1.2.2
  - recommendations: 1.0.0
  - serverinfo: 1.11.0
  - settings: 1.3.0
  - sharebymail: 1.11.0
  - spreed: 11.1.2
  - support: 1.4.0
  - survey_client: 1.9.0
  - systemtags: 1.11.0
  - tasks: 0.13.6
  - text: 3.2.0
  - theming: 1.12.0
  - timemanager: 0.1.8
  - twofactor_backupcodes: 1.10.0
  - twofactor_totp: 6.0.0
  - updatenotification: 1.11.0
  - user_ldap: 1.11.0
  - user_status: 1.1.1
  - user_usage_report: 1.5.0
  - video_converter: 1.0.1
  - viewer: 1.5.0
  - weather_status: 1.1.0
  - workflow_ocr: 1.21.0
  - workflow_pdf_converter: 1.6.0
  - workflowengine: 2.3.0

**The content of config/config.php:**
```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.tthecreator.win"
        ],
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "version": "21.0.1.1",
        "config_is_read_only": false,
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": true
            }
        ],
        "maintenance": false,
        "mysql.utf8mb4": true,
        "supportedDatabases": [
            "mysql"
        ],
        "filesystem_check_changes": 0,
        "debug": false,
        "theme": "",
        "loglevel": 2,
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
        "overwrite.cli.url": "https:\/\/cloud.tthecreator.win",
        "encryption.legacy_format_support": false,
        "default_phone_region": "NL",
        "memcache.local": "\\OC\\Memcache\\APCu"
    }
}
```

#### Client configuration
**Browser:**
Firefox

**Operating system:**
W10
##### Web server error log
N/A

##### Server log (data/nextcloud.log)
```
{"reqId":"QCRVX3O3H3KgutQ7njnS","level":3,"time":"2021-05-04T19:30:18+00:00","remoteAddr":"192.168.2.194","user":"admin_2","app":"index","method":"POST","url":"/index.php/apps/twofactor_totp/settings/enable","message":{"Exception":"Exception","Message":"substr() expects parameter 1 to be string, bool given","Code":0,"Trace":[{"file":"/var/www/html/lib/private/AppFramework/App.php","line":157,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\TwoFactorTOTP\\Controller\\SettingsController"},"enable"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":302,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\TwoFactorTOTP\\Controller\\SettingsController","enable",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"twofactor_totp.settings.enable"}]},{"file":"/var/www/html/lib/base.php","line":993,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/twofactor_totp/settings/enable"]},{"file":"/var/www/html/index.php","line":37,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","Line":159,"Previous":{"Exception":"TypeError","Message":"substr() expects parameter 1 to be string, bool given","Code":0,"Trace":[{"file":"/var/www/html/lib/private/Security/Crypto.php","line":95,"function":"substr","args":[false,0,32]},{"file":"/var/www/html/apps/twofactor_totp/lib/Service/Totp.php","line":95,"function":"encrypt","class":"OC\\Security\\Crypto","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/html/apps/twofactor_totp/lib/Controller/SettingsController.php","line":87,"function":"createSecret","class":"OCA\\TwoFactorTOTP\\Service\\Totp","type":"->","args":[{"__class__":"OC\\User\\User"}]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":218,"function":"enable","class":"OCA\\TwoFactorTOTP\\Controller\\SettingsController","type":"->","args":[1,null]},{"file":"/var/www/html/lib/private/AppFramework/Http/Dispatcher.php","line":127,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\TwoFactorTOTP\\Controller\\SettingsController"},"enable"]},{"file":"/var/www/html/lib/private/AppFramework/App.php","line":157,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OCA\\TwoFactorTOTP\\Controller\\SettingsController"},"enable"]},{"file":"/var/www/html/lib/private/Route/Router.php","line":302,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OCA\\TwoFactorTOTP\\Controller\\SettingsController","enable",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"twofactor_totp.settings.enable"}]},{"file":"/var/www/html/lib/base.php","line":993,"function":"match","class":"OC\\Route\\Router","type":"->","args":["/apps/twofactor_totp/settings/enable"]},{"file":"/var/www/html/index.php","line":37,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"/var/www/html/lib/private/Security/Crypto.php","Line":95},"CustomMessage":"--"},"userAgent":"Mozilla/5.0 (Windows NT 10.0; rv:78.0) Gecko/20100101 Firefox/78.0","version":"21.0.1.1"}
```

##### Browser log
```
XHRPOSThttps://REDACTED/index.php/apps/twofactor_totp/settings/enable
[HTTP/1.1 500 Internal Server Error 449ms]

[ERROR] twofactor_totp: Could not enable TOTP 
Object { app: "twofactor_totp", uid: "admin_2", config: {…}, request: XMLHttpRequest, response: {…}, isAxiosError: true, toJSON: toJSON() }
ConsoleLogger.js:54:18
    value ConsoleLogger.js:54
    value ConsoleLogger.js:80
    createTOTP PersonalTotpSettings.vue:115
    (Async: promise callback)
    createTOTP PersonalTotpSettings.vue:111
    toggleEnabled PersonalTotpSettings.vue:89
    VueJS 3
	{
  "app": "twofactor_totp",
  "uid": "admin_2",
  "config": {
    "url": "/index.php/apps/twofactor_totp/settings/enable",
    "method": "post",
    "data": "{\"state\":1}",
    "headers": {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json;charset=utf-8",
      "requesttoken": "REDACTED"
    },
    "transformRequest": [
      null
    ],
    "transformResponse": [
      null
    ],
    "timeout": 0,
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
    "maxContentLength": -1,
    "maxBodyLength": -1
  },
  "request": {},
  "response": {
    "data": "<!DOCTYPE html>\n<html class=\"ng-csp\" data-placeholder-focus=\"false\" lang=\"en\" data-locale=\"en\" >\n\t<head\n data-requesttoken=\"\">\n\t\t<meta charset=\"utf-8\">\n\t\t<title>\n\t\tNextcloud\t\t</title>\n\t\t<meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n\t\t<meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0, minimum-scale=1.0\">\n\t\t\t\t<meta name=\"apple-itunes-app\" content=\"app-id=1125420102\">\n\t\t\t\t<meta name=\"theme-color\" content=\"#3973C9\">\n\t\t<link rel=\"icon\" href=\"/core/img/favicon.ico\">\n\t\t<link rel=\"apple-touch-icon\" href=\"/core/img/favicon-touch.png\">\n\t\t<link rel=\"mask-icon\" sizes=\"any\" href=\"/core/img/favicon-mask.svg\" color=\"#3973C9\">\n\t\t<link rel=\"manifest\" href=\"/index.php/apps/theming/manifest?v=2\">\n\t\t<link rel=\"stylesheet\" href=\"/core/css/guest.css?v=8b3e0eb3-2\">\n\t\t<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/core/js/dist/main.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/core/js/dist/files_fileinfo.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/core/js/dist/files_client.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/index.php/js/core/merged-template-prepend.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/apps/accessibility/js/accessibilityoca.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/apps/files_sharing/js/dist/main.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/apps/epubreader/js/plugin.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/apps/files_videoplayer/js/main.js?v=8b3e0eb3-2\"></script>\n<script nonce=\"T3BUVVFJUmFiS0JQN0ZSamo3VjNLaVhvdEE0K1FGWE1wZE1LdGI5TCswUT06Q3ZHQkkrWnZEOGQvbWk0NjJPUkhmbWVhd0V4R0FqbWhrTHg4M3RZQXlDMD0=\" defer src=\"/apps/notifications/js/notifications-main.js?v=8b3e0eb3-2\"></script>\n\t\t\t</head>\n\t<body id=\"body-login\">\n\t\t<noscript>\n\t<div id=\"nojavascript\">\n\t\t<div>\n\t\t\tThis application requires JavaScript for correct operation. Please <a href=\"https://www.enable-javascript.com/\" target=\"_blank\" rel=\"noreferrer noopener\">enable JavaScript</a> and reload the page.\t\t</div>\n\t</div>\n</noscript>\n\t\t\t\t\t<input type=\"hidden\" id=\"initial-state-accessibility-data\" value=\"eyJ0aGVtZSI6ZmFsc2UsImhpZ2hjb250cmFzdCI6ZmFsc2V9\">\n\t\t\t\t<div class=\"wrapper\">\n\t\t\t<div class=\"v-align\">\n\t\t\t\t\t\t\t\t\t<header role=\"banner\">\n\t\t\t\t\t\t<div id=\"header\">\n\t\t\t\t\t\t\t<div class=\"logo\">\n\t\t\t\t\t\t\t\t<h1 class=\"hidden-visually\">\n\t\t\t\t\t\t\t\t\tNextcloud\t\t\t\t\t\t\t\t</h1>\n\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t</div>\n\t\t\t\t\t\t</div>\n\t\t\t\t\t</header>\n\t\t\t\t\t\t\t\t<main>\n\t\t\t\t\t<div class=\"error error-wide\">\n\t<h2>Internal Server Error</h2>\n\t<p>The server was unable to complete your request.</p>\n\t<p>If this happens again, please send the technical details below to the server administrator.</p>\n\t<p>More details can be found in the server log.</p>\n\n\t<h3>Technical details</h3>\n\t<ul>\n\t\t<li>Remote Address: 192.168.2.194</li>\n\t\t<li>Request ID: u0eII2sQNDB2FuIBWqvW</li>\n\t\t\t</ul>\n\n\t</div>\n\t\t\t\t</main>\n\t\t\t</div>\n\t\t</div>\n\t\t<footer role=\"contentinfo\">\n\t\t\t<p class=\"info\">\n\t\t\t\t<a href=\"https://nextcloud.com\" target=\"_blank\" rel=\"noreferrer noopener\" class=\"entity-name\">Nextcloud</a> – a safe home for all your data\t\t\t</p>\n\t\t</footer>\n\t</body>\n</html>\n",
    "status": 500,
    "statusText": "Internal Server Error",
    "headers": {
      "cache-control": "no-store, no-cache, must-revalidate",
      "connection": "close",
      "content-length": "4025",
      "content-security-policy": "default-src 'self'; script-src 'self' 'nonce-REDACTED'; style-src 'self' 'unsafe-inline'; frame-src *; img-src * data: blob:; font-src 'self' data:; media-src *; connect-src *; object-src 'none'; base-uri 'self';",
      "content-type": "text/html; charset=UTF-8",
      "date": "Tue, 04 May 2021 19:43:00 GMT",
      "expires": "Thu, 19 Nov 1981 08:52:00 GMT",
      "pragma": "no-cache",
      "referrer-policy": "no-referrer",
      "server": "Apache/2.4.38 (Debian)",
      "strict-transport-security": "max-age=15552000; includeSubDomains",
      "x-content-type-options": "nosniff",
      "x-download-options": "noopen",
      "x-frame-options": "SAMEORIGIN",
      "x-permitted-cross-domain-policies": "none",
      "x-powered-by": "PHP/7.4.18",
      "x-robots-tag": "none",
      "x-xss-protection": "1; mode=block"
    },
    "config": {
      "url": "/index.php/apps/twofactor_totp/settings/enable",
      "method": "post",
      "data": "{\"state\":1}",
      "headers": {
        "Accept": "application/json, text/plain, */*",
        "Content-Type": "application/json;charset=utf-8",
        "requesttoken": "REDACTED"
      },
      "transformRequest": [
        null
      ],
      "transformResponse": [
        null
      ],
      "timeout": 0,
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
      "maxContentLength": -1,
      "maxBodyLength": -1
    },
    "request": {}
  },
  "isAxiosError": true
}
```

**Comment by ChristophWurst** [comment]
``\OC\Security\Crypto::encrypt`` fails because ``hash_hkdf`` returns `false` due to a failure.

> The content of config/config.php

Your config should have a `secret`. This is the source of the problem.

**Comment by Tthecreator** [comment]
Thank you @ChristophWurst !
This was indeed the case.
What I did is take another config from an old install I had and looked at the format of the secret. Then I generated a new one with the same characters and length. Then I couldn't log in actually and I had to remove the data/user/files_encryption/OC_DEFAULT_MODULE file. (I backed it up to be sure). I didn't encrypt any of my data so I'm fine. If anyone looking at this issue did encrypt their data they might be into some more trouble.
